src/meshapi/views/panoramas.py

```diff
@@ -0,0 +1,170 @@
+    token = os.environ.get("PANO_GITHUB_TOKEN")
+
+    if not owner or not repo or not branch or not directory or not host_url or not token:
+        return Response({"detail": "Did not find environment variables"}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    head_tree_sha = get_head_tree_sha(owner, repo, branch)
+    if not head_tree_sha:
+        return Response(
+            {"detail": "Could not get head tree SHA from GitHub"}, status=status.HTTP_500_INTERNAL_SERVER_ERROR
+        )
+
+    panorama_files = list_files_in_git_directory(owner, repo, directory, head_tree_sha)
+    if not panorama_files:
+        return Response({"detail": "Could not get file list from GitHub"}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    print(panorama_files)
+
+    panos = build_pano_dict(panorama_files)
+
+    panoramas_saved, warnings = set_panoramas(panos)
+
+    return Response(
+        {
+            "detail": f"Saved {panoramas_saved} panoramas. Got {len(warnings)} warnings.",
+            "saved": panoramas_saved,
+            "warnings": len(warnings),
+            "warn_install_nums": warnings,
+        },
+        status=status.HTTP_200_OK,
+    )
+
+
+def set_panoramas(panos: dict[str, list[str]]) -> tuple[int, list[str]]:
+    panoramas_saved = 0
+    warnings = []
+
+    host_url = os.environ.get("PANO_HOST")
+
+    for install_number, filenames in panos.items():
+        try:
+            install: Install = Install.objects.get(install_number=int(install_number))
+            panoramas = []
+            if not install:
+                print(
+                    f"Warning: Could not add panorama to building (Install #{install_number}). Install does not exist."
+                )
+                warnings.append(install_number)
+                continue
+            for filename in filenames:
+                file_url = f"{host_url}{filename}"
+                panoramas.append(file_url)
+            if install.building.panoramas == panoramas:
+                continue
+            install.building.panoramas = panoramas
+            install.building.save()
+            panoramas_saved += len(filenames)
+        except Exception as e:
+            print(f"Warning: Could not add panorama to building (Install #{install_number}): {e}")
+            warnings.append(install_number)
+    return panoramas_saved, warnings
+
+
+def build_pano_dict(files: list[str]):
+    panos = {}
+    for f in files:
+        try:
+            number, label = parse_pano_title(Path(f).stem)
+        except BadPanoramaTitle as e:
+            print(e)
+            continue
+        if number not in panos:
+            panos[number] = [f]
+        else:
+            panos[number].append(f)
+    return panos
+
+
+# This is awful. Maybe there are easy ways to generalize some cases like stripping
+# spaces, but for now I would rather explicitly handle these cases until I have
+# better tests.
+def parse_pano_title(title: str):
```

Comment:
Wow this is literally like a leetcode problem I'm so sorry

Reply:
That's what I was thinking the whole time.